### PR TITLE
feat: allow for Pages projects to upload sourcemaps

### PR DESCRIPTION
**Pull request by zebp**
Adds the ability for Pages projects to use the new sourcemap support in Workers to get remapped stacktraces in the dashboard's real-time logs.

Fixes WP-959 WP-960 WP-961.

## Author has addressed the following

- Tests
  - [ ] TODO (before merge)
  - [x] Included
  - [ ] Not necessary because:
- E2E Tests CI Job required? (Use "e2e" label or ask maintainer to run separately)
  - [x] I don't know
  - [ ] Required / Maybe required
  - [ ] Not required because:
- Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))
  - [ ] TODO (before merge)
  - [x] Included
  - [ ] Not necessary because:
- Public documentation
  - [ ] TODO (before merge)
  - [x] Cloudflare docs PR(s): https://github.com/cloudflare/cloudflare-docs/pull/14929

### CODE REVIEW

File: packages/wrangler/src/deployment-bundle/find-additional-modules.ts
Comment (zebp) [inline packages/wrangler/src/deployment-bundle/find-additional-modules.ts:106]: I don't particularly like this solution, but it is functional and minimally invasive. LMK if you think this isn't adequate and I can do some refactoring to make this less hacked-on.

File: .changeset/old-horses-push.md
Comment (penalosa) [inline .changeset/old-horses-push.md:2]:
```suggestion
"wrangler": minor
```

File: packages/wrangler/src/pages/build.ts
Comment (penalosa) [inline packages/wrangler/src/pages/build.ts:238]: There are various other bundling functions in this file that should probably also take this argument (e.g. `buildRawWorker()`)

File: packages/wrangler/src/deployment-bundle/worker.ts
Comment (penalosa) [inline packages/wrangler/src/deployment-bundle/worker.ts:53]: What does "Javascript type" mean? Is this ESM/commonJS?

File: packages/wrangler/src/deployment-bundle/worker.ts
Comment (penalosa) [inline packages/wrangler/src/deployment-bundle/worker.ts:54]:
```suggestion
	 * if we're deploying with sourcemaps enabled. Since we copy extra modules that aren't bundled
```

File: packages/wrangler/src/config/validation-pages.ts
Comment (penalosa) [inline packages/wrangler/src/config/validation-pages.ts:41]: Does this feature require `wrangler.toml`? If so, it's the first Pages feature that doesn't work without a config file—is that okay?

File: packages/wrangler/src/deployment-bundle/source-maps.ts
Comment (penalosa) [inline packages/wrangler/src/deployment-bundle/source-maps.ts:109]: This seems relatively brittle—what if there are two emoty lines?

File: packages/wrangler/src/deployment-bundle/source-maps.ts
Comment (penalosa) [inline packages/wrangler/src/deployment-bundle/source-maps.ts:127]:
```suggestion
	const sourcemapPath = path.join(path.dirname(module.filePath), commentPath);
```

File: packages/wrangler/src/config/validation-pages.ts
Comment (zebp) [reply packages/wrangler/src/config/validation-pages.ts:41]: I've added a `--upload-source-maps` flag for the deploy command and used the existing `--sourcemap` flag for the internal pages function build command. The explicit argument will take precedence over whatever is in the wrangler.toml

File: packages/wrangler/src/__tests__/pages/deploy.test.ts
Comment (penalosa) [inline packages/wrangler/src/__tests__/pages/deploy.test.ts:5270]: These code blocks should all use `dedent` (from `ts-dedent`)

**Review by penalosa — APPROVED** [review]
Pending Frameworks approval. It'd also be great to see an e2e test or similar of this uploading sourcemaps from a framework build process

File: packages/wrangler/src/config/validation-pages.ts
Comment (CarmenPopoviciu) [inline packages/wrangler/src/config/validation-pages.ts:41]: I'm assuming this requires a docs update as well. Pls don't forget to reference it to this PR 🙏

File: packages/wrangler/src/deployment-bundle/source-maps.ts
Comment (dario-piotrowicz) [reply packages/wrangler/src/deployment-bundle/source-maps.ts:114]: PS: ideally I would enter `sourceMapForModule` only for non-virtual modules, but I think that that would be a bit annoying with the `CfModule` type and the fact that we eslint-ban non-null assertions... so this would be fine for me

**Comment by dario-piotrowicz** [comment]
Tested with the Remix sample @zebp graciously provided ❤️ 

I can confirm that the dash realtime logs seem to be picking up the sourcemaps 😄 🥳 

<details>

By introducing an issue in the `_index.tsx` file, without sourcemaps (`wrangler@latest`) I get:
```
    {
      "stack": "    at functionsWorker-0.4868909124392582.js:46912:17\n    at new Promise (<anonymous>)\n    at loader (functionsWorker-0.4868909124392582.js:46911:9)\n    at Object.callRouteLoader (functionsWorker-0.4868909124392582.js:5981:26)\n    at commonRoute.loader (functionsWorker-0.4868909124392582.js:6083:45)\n    at actualHandler (functionsWorker-0.4868909124392582.js:3822:18)\n    at functionsWorker-0.4868909124392582.js:3834:31\n    at runHandler (functionsWorker-0.4868909124392582.js:3845:13)\n    at callLoaderOrAction (functionsWorker-0.4868909124392582.js:3894:26)\n    at Object.resolve (functionsWorker-0.4868909124392582.js:3793:33)",
      "name": "Error",
      "message": "error from _index.tsx",
      "timestamp": 1717496670342
    }
```
whilst with sourcemaps (using this PR's prerelease) I get:
```
   {
      "stack": "    at ../app/routes/_index.tsx:8:13\n    at new Promise (<anonymous>)\n    at loader (../app/routes/_index.tsx:7:5)\n    at loader (../../node_modules/.pnpm/@remix-run+server-runtime@2.9.2_typescript@5.4.5/node_modules/@remix-run/server-runtime/dist/data.js:66:22)\n    at commonRoute.loader (../../node_modules/.pnpm/@remix-run+server-runtime@2.9.2_typescript@5.4.5/node_modules/@remix-run/server-runtime/dist/routes.js:54:39)\n    at handler (../../node_modules/.pnpm/@remix-run+router@1.16.1/node_modules/@remix-run/router/router.ts:4250:14)\n    at ../../node_modules/.pnpm/@remix-run+router@1.16.1/node_modules/@remix-run/router/router.ts:4266:27\n    at runHandler (../../node_modules/.pnpm/@remix-run+router@1.16.1/node_modules/@remix-run/router/router.ts:4271:7)\n    at runHandler (../../node_modules/.pnpm/@remix-run+router@1.16.1/node_modules/@remix-run/router/router.ts:4328:22)\n    at callLoaderOrAction (../../node_modules/.pnpm/@remix-run+router@1.16.1/node_modules/@remix-run/router/router.ts:4179:13)",
      "name": "Error",
      "message": "error from _index.tsx",
      "timestamp": 1717496195315
    }
```

💪 🚀 

</details>

**Comment by dario-piotrowicz** [comment]
The changes generally look good to me, except the few small points I commented before 👆

But also, as it has been pointed out by @CarmenPopoviciu the docs haven't yet been updated
Screenshot 2024-06-04 at 11 32 02
👀 
I don't really know the wrangler's team policy and if this should or not block the PR, I'm leaving this to them 🙂 (@CarmenPopoviciu, @petebacondarwin?) 

Regarding the lack of e2es @penalosa mentioned.... yeah that'd also be quite a nice addition but I can see that the `deploy` function is unit tested... so I guess I am not too too concerned about the lack of an e2e 🤔 (but again I'd really leave this to the wrangler team to decide)

**Comment by penalosa** [comment]
@zebp do you have a docs ticket/PR you can link?

**Comment by zebp** [comment]
> @zebp do you have a docs ticket/PR you can link?

I've started one but I haven't created the PR yet, I'll have it up in a few hours.

File: packages/wrangler/src/config/validation-pages.ts
Comment (zebp) [reply packages/wrangler/src/config/validation-pages.ts:41]: Created a docs PR! https://github.com/cloudflare/cloudflare-docs/pull/14929

**Review by dario-piotrowicz — APPROVED** [review]
Looks good to me! thanks a lot for addressing my comments @zebp 🚀 

(and thanks for the docs PR ❤️)

**Comment by ericmatthys** [comment]
Does this help at all with uploading source maps to Sentry for those using `@cloudflare/pages-plugin-sentry`? That's currently not possible and it makes stacktraces in Sentry more difficult.

**Pull request merged**
